proposals/XXXX-debug.md

```diff
@@ -0,0 +1,120 @@
+
+Using buildtime configuration conditions are an industry-standard approach to handling these needs. They distinguish development behavior from their release counterparts.
+
+The exact nature of conditional build testing varies by language. Swift currently allows you to set custom flags at the command line. 
+
+* `-D <#value#>` (for example, `-D debug`) lets you conditionally compile using the `#if <#value#>` test, such as `#if debug`. There is no way to access these flags for runtime tests and no public way to differentiate "debug" from "release" at runtime. 
+* You can manually set asssert firing conditions by supplying the `-assert-config <#value#>` flag. Supply Debug, Release, Unchecked, or DisableReplacement.
+
+### Defining "Development"
+
+This proposal treats the notion of "development" to mean:
+
+* A compilation or execution environment not meant for general release.
+* Code meant for in-house or beta distribution.
+
+It offers two axes of development conditions:
+
+* It checked for optimized compilation. When code is compiled with optimization flags (-O, -Ounchecked, _specifically excluding -Onone/-Oplayground from these tests_), then the code is "optimized"
+* It checks whether asserts can fire. When an assert is included in the build product, then the code supports assert firing.
+
+These definitions are distinct from any tooling including the "Debug" and "Release" schemes used in IDEs like Xcode. Schemes do not play into the proposed Swift checks outside of their specific inclusion of optimization flags.
+
+## Detail Design
+
+This proposal introduces both run-time and compile-time checks. Here is how each check is designed.
+
+#### Run-Time Checks
+
+The swift standard library includes three hidden helper functions: `_isDebugAssertConfiguration`, `_isReleaseAssertConfiguration`, and `_isFastAssertConfiguration`. 
+
+#### Compile-Time Checks
+
+Compile-time checks use the standard `#if <#condition#>` form and include code on each successful branch. To test for "optimized" conditions, the conditional compilation configuration test must wait until any optimization flag is found and trigger for any flag that is not -Onone or -Oplayground (treated as the equivalent to -Onone). Including a command-line `-O` flag such as `-O` or `-Ounchecked` excludes a "debug" compilation. The "assertion" condition waits slightly longer and then uses the value stored in `Opts.RemoveRuntimeAsserts`.
+
+These build configuration tests allow you to use conditional build configurations to incorporate or exclude code specific to development builds as in the following examples:
+
+```swift
+#if !configuration(optimized)
+    // code for un-optimized builds only
+#endif
+
+#if configuration(assertsWillFire)
```

Comment:
For absolutely no justifiable reason, it would make me happy if the test was `assertive` instead of `assertsWillFire`, as in: 

```
#if configuration(assertive)
```

The camelCase in `assertsWillFire` is bothering me for some reason.

Reply:
@samdeane I am open to the name change. I just led with the most obvious. I'll let the team decide what they want to name it if they act on the proposal. Thank you for the feedback.